server/src/main/java/org/elasticsearch/index/bulk/stats/ShardBulkStats.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.bulk.stats;
+
+import org.elasticsearch.common.ExponentiallyWeightedMovingAverage;
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.metrics.MeanMetric;
+import org.elasticsearch.index.shard.IndexShard;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Internal class that maintains relevant shard bulk statistics / metrics.
+ * @see IndexShard
+ */
+public class ShardBulkStats implements BulkOperationListener {
+
+    private final StatsHolder totalStats = new StatsHolder();
+    private static final double ALPHA = 0.1;
+
+    public BulkStats stats() {
+        return totalStats.stats();
+    }
+
+    @Override
+    public void afterBulk(long shardBulkSizeInBytes, long tookInNanos) {
+        totalStats.totalSizeInBytes.inc(shardBulkSizeInBytes);
+        totalStats.shardBulkMetric.inc(tookInNanos);
+        if (totalStats.timeInMillis == null) {
+            totalStats.timeInMillis = new ExponentiallyWeightedMovingAverage(ALPHA, tookInNanos);
+        } else {
+            totalStats.timeInMillis.addValue(tookInNanos);
+        }
+        if (totalStats.sizeInBytes == null) {
+            totalStats.sizeInBytes = new ExponentiallyWeightedMovingAverage(ALPHA, shardBulkSizeInBytes);
+        } else {
+            totalStats.sizeInBytes.addValue(shardBulkSizeInBytes);
+        }
+    }
+
+    static final class StatsHolder {
+        final MeanMetric shardBulkMetric = new MeanMetric();
+        final CounterMetric totalSizeInBytes = new CounterMetric();
+        ExponentiallyWeightedMovingAverage timeInMillis;
+        ExponentiallyWeightedMovingAverage sizeInBytes;
```

Comment:
I think rather than the `null` check above, it would be better to initialize these to `new ExponentiallyWeightedMovingAverage(ALPHA, 0.0)`. We could then get rid of the null check in `afterBulk(...)` as well as `stats()`

Reply:
@dakrone Thanks for reviewing .
I used to do the same thing here, using initialAvg as 0.0, but when the first bulk request came in, it turned out to be 0.1 * newValue. I think this is not my desired results, so I became an first shard bulk statistics to initialize the EWMA. 
Why do you think of 0.1 * newValue in line with expectations

Reply:
I think I would prefer to have the average slowly build up over a period of requests rather than have the check for every bulk request? I need to benchmark this change prior to merging, so we can leave it as-is and see if the overhead is noticeable.

Reply:
ok, I agree with you. What else do I need to do here? @dakrone

Reply:
I need to benchmark this branch to see whether there is any overhead to the extra stat collecting.

Reply:
I just initialize ewma to new ExponentiallyWeightedMovingAverage(ALPHA, 0.0) and remove  null check in afterBulk(...) as well as stats() !  Does it matter ？